### PR TITLE
Do not eta-expand 0-arg methods. Improve eta-expansion for method values.

### LINKED ISSUE #8299: eta-expansion for method on target that needs implicit conversion
```scala
object Test extends App {
  class C

  implicit class CompatibleC(c: C) {
    def foo(x: Int) = ???
  }

  val c: C = ???
  println(c.foo _)
}
```

```scala
15:44 ~/Projects/Master/sandbox (master)$ s
Test.scala:9: error: missing arguments for method foo in class CompatibleC;
follow this method with `_' if you want to treat it as a partially applied function
  println(c.foo _)
            ^
one error found
```

The error is the same if I manually create an implicit def.

### PR DESCRIPTION
**Pull request by adriaanm**
Insert `()` before considering eta-expansion. A small step towards Scala 3, where you will have to apply `()` explicitly to Scala-defined methods that define an argument list.

Also clean up method value expansion: do it directly, instead of attaching a note for later and adapting again, which fixes scala/bug#8299.

### CODE REVIEW

**Comment by szeiger** [comment]
I think this goes in the right direction and it helps with the overloaded methods in the new collection library that required changes in the compiler codebase. Without `-Xsource:2.14`:

```
[error] /Users/szeiger/code/scala/src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala:558: missing argument list for method mapArgInfo
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `mapArgInfo _` or `mapArgInfo(_)` instead of `mapArgInfo`.
[error]       val argInfos = originalCallsite.argInfos.flatMap(mapArgInfo)
[error]                                                        ^
[error] /Users/szeiger/code/scala/src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala:577: missing argument list for method mapArgInfo
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `mapArgInfo _` or `mapArgInfo(_)` instead of `mapArgInfo`.
[error]       val capturedArgInfos = originalClosureInit.capturedArgInfos.flatMap(mapArgInfo)
[error]                                                                           ^
[error] /Users/szeiger/code/scala/src/compiler/scala/tools/reflect/ToolBoxFactory.scala:230: missing argument list for method makeParam
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `makeParam _` or `makeParam(_)` instead of `makeParam`.
[error]           meth setInfo MethodType(freeTerms.map(makeParam).toList, AnyTpe)
[error]                                                 ^
```

These calls now compile with `-Xsource:2.14` because the methods are monomorphic.

There's still the problem of the dual meaning of "expected type" in overload resolution. When we know that we're typechecking a lambda we can set it to a lub of the parameter types (i.e. a glb of the function type) but for typechecking other constructs that would be wrong (you need the opposite variance). While the more aggressive expansion will attempt to typecheck a method reference as a lambda it doesn't know about the parameter types that are necessary to do that.

This limitation causes a regressions due to the new collection library in cases like this:

```scala
    def iden[T](x: T): T = x
    val v = scala.collection.immutable.BitSet(1,2,3).map(iden)
    val vt: scala.collection.immutable.BitSet = v
```

These are the ones where an explicit `_` is not sufficient as a work-around, you have to write `map(x => iden(x))` to make overload resultion supply the correct parameter types that allow typing of the polymorphic lambda.

**Comment by adriaanm** [comment]
The dependence on syntactic function literals goes away in my
overload_proto branch, btw. Let’s discuss in a chat next week.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -912,50 +894,17 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
-            // can't type with the expected type, as we can't recreate the setup in (3) without calling typed
-            // (note that (3) does not call typed to do the polymorphic type instantiation --
-            //  it is called after the tree has been typed with a polymorphic expected result type)
-            if (hasUndets)
-              instantiate(typed(tree0, mode), mode, pt)
-            else
-              typed(tree0, mode, pt)
-          }
+        // (4.2) apply to empty argument list
+        else if (mt.params.isEmpty && (settings.isScala213 || !isFunctionType(pt))) emptyApplication
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: `isScala213` is true for 2.13+, so should it be `!isScala214` here?

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: The intent is to go the other way: anything < 2.13 needs to check the expected type before inserting the empty apply. From 2.13, we’ll always insert them. I’ll add a comment to clarify

**Comment by szeiger** [comment]
@adriaanm Should we merge this for M4 or hold off until you're done with the additional changes in overload_proto?

**Comment by adriaanm** [comment]
I would like to merge it. Lukas hasn't officially signed off on it yet, though. Maybe @retronym can take a look?

**Comment by adriaanm** [comment]
ping @lrytz

**Comment by SethTisue** [comment]
fixes https://github.com/scala/scala-dev/issues/391

**Comment by SethTisue** [comment]
@adriaanm do you want a community build run on this...?

**Comment by adriaanm** [comment]
Hadn't thought about it, but I guess it wouldn't hurt :-)

**Review by lrytz — APPROVED** [review]
LGTM. Push-`-f`'d to your branch, added some tests (diff https://gist.github.com/lrytz/82e301f22610ce7d1b604a069db8b2fc)

**Comment by adriaanm** [comment]
Thanks!

**Comment by SethTisue** [comment]
(I'll keep an eye out for related breakage in future community build runs.)

**Comment by jrudolph** [comment]
Is there any more documentation about why this change was done? Is it for implementation reasons? Or is it a temporary thing needed for an incremental improvement to saner defaults in the upcoming versions?

I see that it is somehow related to 0-arg methods still being allowed to be called without parentheses (why?). Why not change that behavior and keep eta-expansion intact? It's an incompatible change anyway.

While trying to migrate code to 2.13 (in https://github.com/akka/akka-http/pull/2298/files#diff-10c929125531820c5fea3db482903b3eR41) we now have the situation where for

```scala
def f0(): Int = 23
def f1(a: Int): Int = 42
def g(f0: () => Int, f1: Int => Int): Int = ???
```

in 2.12 we have

```scala
g(f0, f1)
```

while for 2.13 a valid replacement is

```scala
g(() => f0, f1)
```

which seems odd. First, it's now inconsistent with the 1-arg case and also the `f0` application still doesn't require parentheses.

Of course, it's not the only possible replacement. `g(f0 _, f1)` might still work but it seems method values are also being deprecated in the future?

**Comment by adriaanm** [comment]
Our assumption is that relying on automatic insertion of `()` is more common than eta-expansion of a 0-arg method, and thus changing the latter affects fewer users than the former.

**Comment by adriaanm** [comment]
oh, wait a minute, looks like I forgot about 

> If the expected type is of the form () => T, we eta expand.

(from https://github.com/lampepfl/dotty/issues/2570#issuecomment-306202339)

So, your example should actually work.

**Comment by jrudolph** [comment]
Hmm, isn't it the same as https://github.com/scala/scala/pull/6475/files#diff-97f418d5b560997cbbea454143f963cfR9 ? Or should that only happen under `-Xsource:2.14`?

```scala
Welcome to Scala 2.13.0-M5 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_181).
Type in expressions for evaluation. Or try :help.

scala> def f0(): Int = 42
f0: ()Int

scala> def g(f0: () => Int) = ???
g: (f0: () => Int)Nothing

scala> g(f0)
         ^
       error: type mismatch;
        found   : Int
        required: () => Int
```

**Comment by adriaanm** [comment]
I was confused (well, we did go back and forth on this a few times), but the latest decision was to allow eta expansion if the expected type demands it. I'll create a follow up PR today

**Comment by jrudolph** [comment]
> I was confused (well, we did go back and forth on this a few times), but the latest decision was to allow eta expansion if the expected type demands it. I'll create a follow up PR today

But is it still possible given the spec update to do Empty Application before eta expansion? It seems the M5 behavior is at least following the spec (even if that's not what I would like).

**Comment by adriaanm** [comment]
Hmm. 🤔 Now that I've refreshed my recollection of the motivation (see especially scala/bug#9178), I wonder whether a type-driven exception is a good idea. For dotty alignment, we'd have to do it, but the puzzlers are puzzling.

> scala/bug#7187 deprecates eta-expansion of zero-arg method values (since 2.12; was never introduced for SAM types).
> The next step is to also deprecate insertion of `()` (except for java-defined methods), as dotty already requires explicitly writing them.
> Once explicit application to () is required, we can more aggressively eta-expand method references, even if they are 0-arity

**Comment by jrudolph** [comment]
Yep, this issue (like so many others) seems to have connections and inter-dependencies with so many other things it's quite hard to keep track of motivations and current progress...

**Comment by adriaanm** [comment]
I just checked the status quo in dotty (https://github.com/dotty-staging/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/Typer.scala#L2491), and it looks like 0-ary eta expansion is performed regardless of expected type (well, there's a  warning when the SAM type is not annotated as FunctionalInterface). I just happened to be in the neighbourhood yesterday but missed this. /cc @odersky @smarter

**Comment by adriaanm** [comment]
Since this issue is spread out over so many tickets, here's the motivation for deprecating eta-expansion in this case:

```scala
def mkStub(): () => String = () => ""
val stub: () => Any = mkStub

println(stub()) // <function0>
```

**Comment by adriaanm** [comment]
I think you could just as well argue that it doesn't make sense to define `mkStub` with an empty argument list, since it's pure in the sense that it doesn't mutate any potentially aliased objects. 

Since you did define mkStub to take an argument list, you should also supply it on (intended) application.

**Comment by adriaanm** [comment]
@jrudolph do you have a real world example in akka or something? In the abstract, it's hard to decide.

**Comment by dwijnand** [comment]
https://github.com/akka/akka-http/pull/2298/files#diff-10c929125531820c5fea3db482903b3eR41 ?

**Comment by jrudolph** [comment]
> @jrudolph do you have a real world example in akka or something? In the abstract, it's hard to decide.

Yes, one instance is in https://github.com/akka/akka-http/pull/2298/files#diff-10c929125531820c5fea3db482903b3eR41. But I think it's hard to discuss concretely because every single case will be quite insignificant when there's a simple workaround like here. I just stumbled over it because in that example we have both Function1 and Function0 next to each other and it shows how the new behavior isn't consistent any more.

It was

```scala
def encodeChunk(bytes: ByteString): ByteString = ...
def finish(): ByteString = ...

StreamUtils.byteStringTransformer(encodeChunk, finish)
```

Now it has to be

```scala
StreamUtils.byteStringTransformer(encodeChunk, () => finish())
```

> Since this issue is spread out over so many tickets, here's the motivation for deprecating eta-expansion in this case:
> 
> ```scala
> def mkStub(): () => String = () => ""
> val stub: () => Any = mkStub
> 
> println(stub()) // <function0>
> ```

In my opinion, this example works as intended. Empty application seems much more like a convenience feature than eta-expansion (but your mileage may vary) so I'd wish for eta-expansion to take precedence. You could go further and say why would you create a 0-arg method at all in Scala if not to mean it to be side-effecting? In that case empty application seems downright dangerous :fire:

**Comment by adriaanm** [comment]
I tend to agree. I think our (future) rule for auto-application should be even stricter than it is in dotty right now. You only get a free `()` when invoking a Java-defined method, otherwise, if the method is defined with an argument list, you must supply it. A Scala method overriding a Java method also gets an exemption: it may drop the empty argument list defined in Java (because you can't express no-argument list in Java), and then any (statically known) application of that argument-list-free method of course needn't supply the empty argument list.

I think we should deprecate along these lines in 2.14, and enforce in 3.0.

**Comment by jrudolph** [comment]
:+1: sounds like a good plan for me.

I think the conflict will then still remains (and a different solution might be chosen) for calling Java methods or creating SAMs of Java interfaces like the one shown in https://github.com/scala/bug/issues/9178.

**Comment by adriaanm** [comment]
See #7660

**Comment by SethTisue** [comment]
I'm not sure if I grasp the whole picture here, but this part for sure:

> You only get a free () when invoking a Java-defined method

is 💯, the way it is now has been bugging me for a decade

**Pull request merged**